synapse/handlers/sliding_sync.py

```diff
@@ -0,0 +1,611 @@
+#
+# This file is licensed under the Affero General Public License (AGPL) version 3.
+#
+# Copyright (C) 2024 New Vector, Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# See the GNU Affero General Public License for more details:
+# <https://www.gnu.org/licenses/agpl-3.0.html>.
+#
+# Originally licensed under the Apache License, Version 2.0:
+# <http://www.apache.org/licenses/LICENSE-2.0>.
+#
+# [This file includes modifications made by New Vector Limited]
+#
+#
+import logging
+from enum import Enum
+from typing import TYPE_CHECKING, AbstractSet, Dict, Final, List, Optional, Tuple
+
+import attr
+from immutabledict import immutabledict
+
+from synapse._pydantic_compat import HAS_PYDANTIC_V2
+
+if TYPE_CHECKING or HAS_PYDANTIC_V2:
+    from pydantic.v1 import Extra
+else:
+    from pydantic import Extra
+
+from synapse.api.constants import Membership
+from synapse.events import EventBase
+from synapse.rest.client.models import SlidingSyncBody
+from synapse.types import JsonMapping, Requester, RoomStreamToken, StreamToken, UserID
+
+if TYPE_CHECKING:
+    from synapse.server import HomeServer
+
+logger = logging.getLogger(__name__)
+
+
+def filter_membership_for_sync(*, membership: str, user_id: str, sender: str) -> bool:
+    """
+    Returns True if the membership event should be included in the sync response,
+    otherwise False.
+
+    Attributes:
+        membership: The membership state of the user in the room.
+        user_id: The user ID that the membership applies to
+        sender: The person who sent the membership event
+    """
+
+    return (
+        # Everything except `Membership.LEAVE` because we want everything that's *still*
+        # relevant to the user. There are few more things to include in the sync response
+        # (newly_left) but those are handled separately.
+        membership in (Membership.LIST - {Membership.LEAVE})
+        # Include kicks
+        or (membership == Membership.LEAVE and sender != user_id)
```

Comment:
I think this can be simplified to:

```suggestion
        membership != Membership.LEAVE or sender != user_id
```

Reply:
I think this is correct but maybe too clever if I ever come back and try to deduce what it's trying to do. A comment will help but the code is a bit abstract for what's trying to accomplish.

Reply:
I was mostly thinking out getting rid of the unnecessary set constructions there. You could also do:

```python
# Filter out plain leaves
not (membership == LEAVE and sender == user_id)
```

Reply:
:+1: Went with your first suggestion and I think the comment turned out well -> https://github.com/element-hq/synapse/pull/17187/commits/6f10b9722df7bad0a32cf0544b93b5940900bb28